proposals/2024-04-10-native-support-for-info-metrics-metadata.md

```diff
@@ -0,0 +1,112 @@
+
+### Pitfalls of the current solution
+
+Prometheus currently persists info metrics as if they were normal float samples.
+This means that knowledge of info metrics' identifying labels are lost, and you have to base yourself on convention when querying them (for example that `target_info` should have `job` and `instance` as identifying labels).
+There's also no particular support for enriching query results with info metric labels in PromQL.
+The consequence is that you need relatively expert level PromQL knowledge to include info metric labels in your query results; as OTel grows in popularity, this becomes more and more of a problem as users will want to include certain labels from `target_info` (corresponding to OTel resource attributes).
+Without persisted info metric metadata, one can't build more user friendly abstractions (e.g. a PromQL function) for including OTel resource attributes (or other info metric labels) in query results.
+Neither can you build dedicated UI for OTel resource attributes (or other info metric labels).
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Persist info metrics with labels categorized as either identifying or non-identifying.
+* Track when info metrics' set of identifying labels changes. This shouldn't be a frequent occurrence, but it should be handled.
+* Automatically treat the old version of an info metric as stale for query result enriching purposes, when its data labels change (producing a new time series, but with same identity).
+* Add TSDB API for, given a certain time series and a certain timestamp, getting data labels, potentially filtered by certain matchers, from info metrics with identifying labels in common with the time series in question.
+* Simplify enriching of query results with info metric labels in PromQL, e.g. via a new function.
+
+### Audience
+
+Prometheus maintainers.
+
+## Non-Goals
+
+## How
+
+* Introduce a new info metric sample type, to track the info metric's identifying label set over time (in case it changes).
+* Augment the head and block indexes with indexes of info metrics, for easy finding of info metrics matching time series.
+* Add a method to the TSDB API for matching info metric data labels to a time series, given a certain timestamp and potentially data label matchers - the method will use the aforementioned head and block info metric indexes.
+* Simplify the inclusion of info metric labels in PromQL through a new `info` function: `info(v instant-vector[, ls label-selector])`.
+  This function will be UI for the aforementioned TSDB API.
+
+Using the `info` function, we can simplify the previously given PromQL join example as follows:
+
+```
+sum by (k8s_cluster_name, http_status_code) (
+  info(
+    rate(http_server_request_duration_seconds_count[2m]),
+    {k8s_cluster_name=~".+"}
+  )
+)
```

Comment:
Can you help me understand this example? I know the current rule for aggregating rates is [Rate then sum, never sum then rate](https://www.robustperception.io/rate-then-sum-never-sum-then-rate).

Now, will this turn into `rate then info then sum`, or could you also `info then rate then sum`?

Not sure if there is any specific reason to do `rate()` before `info()`, but `sum()` after `info()`.

Reply:
Disclaimer: I just copied this example from Beorn's [original document](https://docs.google.com/document/d/1ebhGNLs3uhdeprJCullM-ywA9iMRDg_mmnuFAQCloqY/edit?pli=1#heading=h.37hgysjhge82). I'll answer based on my knowledge of `info` though.

`rate(http_server_request_duration_seconds_count[2m])` becomes a sub-query to the `info` function, which adds the `k8s_cluster_name` label from `target_info` to the `http_server_request_duration_seconds_count` time series resulting from the sub-query. The `sum` expression aggregates by the labels `k8s_cluster_name` and `http_status_code`. If you don't apply `sum` after `info`, the `k8s_cluster_name` label won't be available.

As for whether you could do `info` before `rate`; I haven't tested, but I think it would work. However, it looks less logical/comprehensible to me, as you really want the rate of `http_server_request_duration_seconds_count[2m]`.